File: src/main/java/com/raychenon/leetcode/SumofAbsoluteDifferencesinaSortedArray.java
package com.raychenon.leetcode;

import java.util.Arrays;

/**
 * User: raychenon
 * Date: 18/12/2020
 * https://leetcode.com/problems/sum-of-absolute-differences-in-a-sorted-array/discuss/969761/Illustrations-for-O(n)-formula-Java-Kotlin-Python
 */
public class SumofAbsoluteDifferencesinaSortedArray {

    /**
     * Time complexity: O(n)
     * Space complexity: O(1) , the output space is not counted
     * @param nums
     * @return
     */
    public int[] getSumAbsoluteDifferences(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        int sumBelow = 0;
        int sumTotal = Arrays.stream(nums).sum();

        for (int i = 0; i < n; i++) {
            int num = nums[i];
            sumTotal -= num;
            res[i] = sumTotal - (n - i - 1) * num + i * num - sumBelow;
            sumBelow += num;
        }
        return res;
    }
}

File: src/main/kotlin/com/raychenon/kotlin/leetcode/SumofAbsoluteDifferencesinaSortedArray.kt
package com.raychenon.kotlin.leetcode

/**
 * User: raychenon
 * Date: 18/12/2020
 * https://leetcode.com/problems/sum-of-absolute-differences-in-a-sorted-array/discuss/969761/Illustrations-for-O(n)-formula-Java-Kotlin-Python
 */
object SumofAbsoluteDifferencesinaSortedArray {

    /**
     * Time complexity: O(n)
     * Space complexity: O(1) , the output space is not counted
     * @param nums
     * @return
     */
    fun getSumAbsoluteDifferences(nums: IntArray): IntArray {
        val n = nums.size
        val res = IntArray(n)
        var sumBelow = 0
        var sumTotal = 0
        for (num in nums) {
            sumTotal += num
        }
        for (i in 0 until n) {
            val num = nums[i]
            sumTotal -= num
            res[i] = sumTotal - (n - i - 1) * num + i * num - sumBelow
            sumBelow += num
        }
        return res
    }
}